plugin/workspace_symbol.py

```diff
@@ -0,0 +1,94 @@
+import sublime_plugin
+import sublime
+import time
+from .core.protocol import Request
+from .core.registry import client_for_view, LspTextCommand
+from .core.url import uri_to_filename
+from .symbols import format_symbol_kind
+import os
+
+try:
+    from typing import List, Optional, Dict, Any
+    assert List and Optional and Dict and Any
+except ImportError:
+    pass
+
+
+class Symbol(sublime_plugin.ListInputHandler):
+    def __init__(self, symbols):
+        self._symbols = symbols
+        self._items = list(map(lambda s: (self._format(s), s), self._symbols))
+
+    def _format(self, s: 'Dict[str, Any]') -> str:
+        file_name = os.path.basename(s['location']['uri'])
+        name = "{} - {} -- {}".format(s['name'], s.get('containerName', ""), file_name)
+        return name
+
+    def list_items(self):
+        return self._items
+
+    def preview(self, s):
+        return format_symbol_kind(s["kind"])
+
+
+class QueryHandler(sublime_plugin.TextInputHandler):
+    def __init__(self, view):
+        self.view = view
+        self.client = client_for_view(self.view)
+        self._symbols = []  # type: 'List[Dict]'
+        self._completed_request = False
+        self._request_time = 0
+
+    def _timeout(self) -> bool:
+        return self._request_time > 3
+
+    def _handle_response(self, symbols: 'List[Dict[str, Any]]'):
+        self._symbols = symbols
+        self._completed_request = True
+
+    def _handle_error(self, error: 'Dict[str, Any]') -> None:
+        reason = error.get("message", "none provided by server :(")
+        msg = "command 'workspace/symbol' failed. Reason: {}".format(reason)
+        self._completed_request = True
+        sublime.error_message(msg)
+
+    def validate(self, txt):
+        if txt != "":
+            self._completed_request = False
+            self._request_time = 0
+            request = Request.workspaceSymbol({"query": txt})
+            if self.client:
+                self.client.send_request(request, self._handle_response, self._handle_error)
+
+            # wait for request to complete
+            while not self._completed_request or not self._timeout:
```

Comment:
Sublime UI features do not accommodate async tasks, but holding the UI thread hostage with a time.sleep might result in complaints about freezes.

It's a nice feature that the text and list input handlers can be chained using `next_input`, but I'd prefer that we simply fire the request and let the response handler run a new command with the results provided to a ListInputHandler.

I hope we can remove the timeout, no idea if 3 seconds is a reasonable limit - we want users to feedback on the server to improve the behaviour :)

Reply:
I agree here and have serious doubts about the user experience.

Can't we use something like a `view.set_status` while the request is in-flight, and then clear the `view.set_status` when the request handler gets invoked? And then show a quick_panel?

Reply:
* The reason I combined the two `InputListener` with a wait in the middle is that it could be that your  search query doesn't lead to the results you want (it happens a lot to me). So you want to delete your input (query string) and issue a request again, like in the demo.
If we hide the input field and show a quick panel after and you don't like the results you have to first cancel the quick panel and then invoke the command again. This IMO breaks the user experience.
All this is a workaround because sublime doesn't have `on_change` to make the user experience interactive like in VS code :(

* during my testing using [metals](https://scalameta.org/metals/), the response time was never over a second. very often it was betwee 0.25 and 0.6. So in term of user experience, you don't actually see the UI freeze because the waiting time is very short and you are waiting for the list to populate anyway.
@predragnikolic Maybe you can tell us what was the response time for you with `rls` ?

I can reduce the timeout time but if you guys are still against the current design I can implement what Tom suggested.

Reply:
>I'd prefer that we simply fire the request and let the response handler run a new command with the results provided to a ListInputHandler.

@tomv564 I tried your suggestion just to see the user experience and it doesn't work :/
when running the second command, sublime doesn't show the  `ListInputHandler` containing the list of symbols and goes straight to the the `run` method. 
The `input(args)` method returning the `ListInputHandler` is taken into consideration only if the command is run via the command palette, which is not the case here ...

Reply:
Sorry for the late reply, I tested this with the `lsp-tsserver` in a small js project. 
```
ended
3.3723504543304443   <-- this was the first request
ended
0.05594778060913086
ended
0.04345417022705078
ended
0.19172906875610352
ended
0.025737762451171875
ended
0.11135029792785645
ended
0.05542397499084473
```

Regarding rust. I am just learning rust, I usually experiment and just declare a couple of functions. 
These are the result for the `rls. I had 3 functions defined :D
```
ended
0.0030100345611572266
ended
0.0070781707763671875
ended
0.009324789047241211
```

Reply:
@predragnikolic thanks for the insights. I changed the implementation to avoid the thread sleep now

Reply:
I would avoid hardcoded timeouts if possible, or if you're gonna keep the timeout I recommend to least increase it to something like 20s. If a language server is super slow then that's an issue of the language server, the LSP package is not to blame. The exact response time depends on several factors: the query itself, number of results, and the size of the project, whether it's the first request that triggers indexing, computer hardware, operating system (I/O is noticeably slower on Windows), ...

In Metals we search both in project sources and library dependencies so a query can a lot of results in some cases leading to ~2s response times based on my experiments https://scalameta.org/metals/blog/2019/01/22/bloom-filters.html#response-times I'm sure we can optimize some parts down the road, maybe truncate the number of results for generic queries like `"S"` but at the moment I can't guarantee all responses will be <3s (esp. not for large codebases on Windows machines).

Reply:
thanks for the detailed description. I removed time out now because the results panel is shown asynchronously.